### PR TITLE
feat: add mock contracts for ERC20 and ERC721

### PR DESCRIPTION
**Pull request by andreivladbrg**
Closes https://github.com/foundry-rs/forge-std/issues/469

### CODE REVIEW

**Comment by mds1** [comment]
Thanks for this! Only skimmed the code but lgtm so far, will take a closer look once we get CI passing.

It seems we'll need to make a few changes so it compiles with older versions, see the run [here](https://github.com/foundry-rs/forge-std/actions/runs/6552104405/job/17986006102?pr=470) or the screenshot below. The immutable fix is easy, we can just remove that

The chain ID is a bit trickier but we found a good way to do that in `StdCheats.sol`, which you can copy. In the code you'd call the `_pureChainId()` method: https://github.com/Tudmotu/forge-std/blob/5dcccff9d9fd2ea1b2ef540308313ed31377a982/src/StdCheats.sol#L615-L636


For the constructor, I'm not sure if there's a solution offhand: We want to make sure there's no warnings regardless of the solc version used (you can of course test this easily by changing the `solc_version` in the config file), and my guess is that adding the `public` visibility modifier to the constructor for 0.6.12 will result in a warning for later solc versions. @mattsse if we used `ignored_error_codes` in forge-std's `foundry.toml` is that respected in upstream repos without affecting them? In other words, if we silence warning code 4937 in forge-std, does that both (1) hide forge-std's emission of that warning from upstream users, and (2) not hide upstream user's own emission of that warning? If so, that might be a suitable solution. If not, we might have to have two versions of each contract

image

**Comment by andreivladbrg** [comment]
@mds1, just wanted to let you know that I've pushed a commit to fix the incompatibilities with older solc versions (there were more 😁).

Regarding the constructor, it's indeed a tricky one. I'm not entirely sure how to address it yet. For now, I've set the `public` modifier, but there is a warning when compiling (tho it runs).

Another idea, aside from having two versions of the contracts, is to remove the constructor. We could then hard-code the metadata to default values like "Mock ERC20", "MOCK", decimals to 18, and add a `changeMetadata` method that allows adjustments to these values later on.

wdyt?

**Comment by mds1** [comment]
Here's a constructor workaround we used to use in forge-std which should work here

```solidity
/// @dev To hide constructor warnings across solc versions due to different constructor visibility requirements and
/// syntaxes, we put the constructor in a private method and assign an unused return value to a variable. This
/// forces the method to run during construction, but without declaring an explicit constructor.
uint256 private CONSTRUCTOR = _constructor();

function _constructor() private returns (uint256) {
  // logic here
  return 0;
}
```

**Comment by andreivladbrg** [comment]
> Here's a constructor workaround we used to use in forge-std which should work here

It is a smart way of doing this, but I don't think it is entirely useful here, given that we also need to pass arguments in the constructor (name, symbol).

>remove the constructor and have an initialize method that can only be called once, like proxies use. Then in StdUtils we can add deployMockERC20 and deployMockERC721 helper functions to simplify the deploy + initialization process.

I ended up doing this way.

File: src/StdUtils.sol [hunk 1]
```diff
@@ -133,6 +135,21 @@ abstract contract StdUtils {
         return computeCreate2Address(salt, initCodeHash, CREATE2_FACTORY);
     }
 
+    /// @dev returns a initilized mock ERC20 contract
```
Comment (mds1) [inline src/StdUtils.sol:138]:
```suggestion
    /// @dev returns an initialized mock ERC20 contract
```

File: src/StdUtils.sol [hunk 2]
```diff
@@ -133,6 +135,21 @@ abstract contract StdUtils {
...
+    function deployERC20Mock(string memory name, string memory symbol, uint8 decimals)
+        internal
+        returns (ERC20Mock mock)
+    {
+        mock = new ERC20Mock();
+        mock.initialize(name, symbol, decimals);
+    }
+
+    /// @dev returns a initilized mock ERC721 contract
```
Comment (mds1) [inline src/StdUtils.sol:147]:
```suggestion
    /// @dev returns an initialized mock ERC721 contract
```

File: src/StdUtils.sol [hunk 3]
```diff
@@ -133,6 +135,21 @@ abstract contract StdUtils {
         return computeCreate2Address(salt, initCodeHash, CREATE2_FACTORY);
     }
 
+    /// @dev returns a initilized mock ERC20 contract
+    function deployERC20Mock(string memory name, string memory symbol, uint8 decimals)
+        internal
+        returns (ERC20Mock mock)
+    {
+        mock = new ERC20Mock();
```
Comment (mds1) [inline src/StdUtils.sol:143]: Let's rename everything to put the "Mock" part as a prefix, e.g. `MockERC20` and `MockERC721`—this is the naming convention solmate uses, and it's my preference as well. For example:
```suggestion
    function deployMockERC20(string memory name, string memory symbol, uint8 decimals)
        internal
        returns (MockERC20 mock)
    {
        mock = new MockERC20();
```

File: src/mocks/ERC20Mock.sol [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+
+    /*//////////////////////////////////////////////////////////////
+                               INITIALIZE
+    //////////////////////////////////////////////////////////////*/
+
+    /// @dev A bool to track whether the contract has been initialized.
+    bool private initialized;
+
+    /// @dev To hide constructor warnings across solc versions due to different constructor visibility requirements and
+    /// syntaxes, we add a initialization function that can be called only once.
```
Comment (mds1) [inline src/mocks/ERC20Mock.sol:53]:
```suggestion
    /// syntaxes, we add an initialization function that can be called only once.
```

File: src/mocks/ERC721Mock.sol [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+
+    /*//////////////////////////////////////////////////////////////
+                               INITIALIZE
+    //////////////////////////////////////////////////////////////*/
+
+    /// @dev A bool to track whether the contract has been initialized.
+    bool private initialized;
+
+    /// @dev To hide constructor warnings across solc versions due to different constructor visibility requirements and
+    /// syntaxes, we add a initialization function that can be called only once.
```
Comment (mds1) [inline src/mocks/ERC721Mock.sol:61]:
```suggestion
    /// syntaxes, we add an initialization function that can be called only once.
```

File: src/mocks/ERC20Mock.sol
Comment (mds1) [inline src/mocks/ERC20Mock.sol:1]: Would you mind adding the test suites from Solmate also and using those here? They're pretty thorough, and from a code review of your contracts it seems to correct, but having tests would add confidence + protect us from introducing any bugs in future PRs/changes to this code

**Review by mds1 — COMMENTED** [review]
Good call about the constructor, `initialize` approach used here LGTM

File: src/mocks/ERC20Mock.sol
Comment (andreivladbrg) [reply src/mocks/ERC20Mock.sol:1]: > Would you mind adding the test suites from Solmate also and using those here 

Sure, just added them

> protect us from introducing any bugs in future PRs/changes to this code

Correct

**Comment by andreivladbrg** [comment]
@mds1 Addressed all your comments, lmk if there is anything else

**Review by mds1 — APPROVED** [review]
This is great, thank you!

Would you mind adding a PR to the book's forge standard library reference documenting these mocks and deploy utils?

**Pull request merged**